ncc/examples/basic.c

```diff
@@ -0,0 +1,1927 @@
+
+u64** vm_commands_selected = NULL;
+u64** vm_commands_root = NULL;
+
+// Vector containing variable values
+size_t vm_vars_allocated = 0;
+i64 vm_vars[1024];
+
+// Stack that can be pushed to and poped from
+size_t vm_stack_cursor = 0;
+i64 vm_stack[1024];
+
+u64** vm_commands_alloc(u64 num)
+{
+    size_t capacity = 256;
+    size_t meta_data_size = sizeof(u64*)*5;
+    size_t inst_buffer_size = sizeof(u64)*capacity;
+
+    u64** command_meta = (u64**)malloc(meta_data_size);
+    NULLGAURD(command_meta);
+    memset(command_meta, 0, meta_data_size);
+    u64* inst_buffer = (u64*) malloc(inst_buffer_size);
+    NULLGAURD(inst_buffer);
+    memset(inst_buffer, 0, inst_buffer_size);
+    command_meta[VM_COMMANDS_CAPACITY] = (u64*)capacity;
+    command_meta[VM_COMMANDS_NUM] = (u64*)num;
+    command_meta[VM_COMMANDS_INSTS_BUFFER] = inst_buffer;
+    return command_meta;
+}
+
+u64** vm_command_find(u64 num)
+{
+    for(u64** cur_cmd = vm_commands_root; cur_cmd != 0; cur_cmd = (u64**)cur_cmd[VM_COMMANDS_NEXT])
+    {
+        u64 cur_cmd_num = (u64)cur_cmd[VM_COMMANDS_NUM];
+        if (cur_cmd_num == num) return cur_cmd;
+    }
+    return 0;
+}
+
+void vm_command_free(u64** cmd)
+{
+    free((void*)cmd[VM_COMMANDS_INSTS_BUFFER]);
+    free((void*) cmd);
+}
+
+u64** vm_command_create(u64 num)
+{
+
+    u64** new_cmd = vm_commands_alloc(num);
+    NULLGAURD(new_cmd);
+
+    if(vm_commands_root == NULL)
+    { // initialize
+        DEBUG("Init commands\n");
+        new_cmd[VM_COMMANDS_NEXT] = (u64*)vm_commands_root;
+        vm_commands_root = new_cmd;
+        return new_cmd;
+    } 
+
+    DEBUG("Finding command\n");
+    u64** prev_cmd = 0;
+    for(u64** cur_cmd = vm_commands_root;; cur_cmd = (u64**)cur_cmd[VM_COMMANDS_NEXT])
+    {
+        if(cur_cmd == NULL)
+        {
+            DEBUG("Inserting command at the end");
+            prev_cmd[VM_COMMANDS_NEXT] = (u64*)new_cmd;
+            break;
+        }
+        u64 cur_cmd_num = (u64)cur_cmd[VM_COMMANDS_NUM];
+        if(cur_cmd_num > num)
+        { // insert command before
+            DEBUG("Inserting commands in between commands");
+            if(cur_cmd == vm_commands_root) vm_commands_root = new_cmd;
+            prev_cmd[VM_COMMANDS_NEXT] = (u64*)new_cmd;
+            new_cmd[VM_COMMANDS_NEXT] = (u64*)cur_cmd;
+            break;
+        }
+        else if(cur_cmd_num == num)
+        { // insert command exactly at
+            DEBUG("Replacing existing command");
+            if(cur_cmd == vm_commands_root) vm_commands_root = new_cmd;
+            prev_cmd[VM_COMMANDS_NEXT] = (u64*)new_cmd;
+            new_cmd[VM_COMMANDS_NEXT] = (u64*)cur_cmd[VM_COMMANDS_NEXT];
+            vm_command_free(cur_cmd);
+            break;
+        }
+        prev_cmd = cur_cmd;
+    }
+
+    return new_cmd;
+}
+
+u64 vm_alloc_var()
+{
+    u64 allocated = vm_vars_allocated;
+    ++vm_vars_allocated;
+    return allocated;
+}
+
+int strcmp_with_len(char* a, char* b, size_t b_len)
```

Comment:
I believe this is equivalent to `strncmp`: https://cplusplus.com/reference/cstring/strncmp/

We should probably implement this function in `include/string.h` and use that.